### PR TITLE
WIP: Issue 468 member's account deletion

### PR DESCRIPTION
**Pull request by cesy**
Fixes issue #468

Not ready to merge yet.

Please refer to previous pull request (PR #662) for discussion.

### CODE REVIEW

**Comment by homu-gs** [comment]
:umbrella: The latest upstream changes (presumably #1059) made this pull request unmergeable. Please resolve the merge conflicts.

File: app/controllers/admin_controller.rb [hunk 1]
```diff
@@ -8,7 +8,7 @@ def index
 
   def newsletter
     authorize! :manage, :all
-    @members = Member.confirmed.wants_newsletter.all
+    @members = Member.confirmed.undeleted.wants_newsletter.all
```
Comment (Br3nda) [inline app/controllers/admin_controller.rb]: Could we use a default scope instead? All queries become undeleted members, instead of changing all queries.

Comment (Br3nda) [inline app/controllers/admin_controller.rb]: example here
http://api.rubyonrails.org/classes/ActiveRecord/Scoping/Default/ClassMethods.html

```
class Member < ActiveRecord::Base
  def self.default_scope
    where(deleted: false)
  end
end

Member.all                                  # Fires "SELECT * FROM member WHERE deleted = false"
Member.unscoped.all                         # Fires "SELECT * FROM members"
Member.where(deleted: true).unscoped.all # Fires "SELECT * FROM members"
```

Comment (CloCkWeRX) [inline app/controllers/admin_controller.rb]: Default scopes tend to make assumptions for one context, at the cost of others and aren't super obvious all of the time. Undeleted... yeah ok, it would make a lot of sense; but it can be tempting to add in default sorts or other bits that are web UI centric. Same sort of scope creep as before hooks can sneak in.

If we do add it, lets prefer a light touch :)

**Comment by maco** [comment]
I like the default scope idea. We should never be exposing deleted users
over web, API, or anything else. The deleted ones should basically only be
operated on by migrations.

**Comment by pozorvlak** [comment]
What @maco said :-)

**Comment by maco** [comment]
@cesy what do you think about the default scope idea?

**Comment by maco** [comment]
(Given the recent spam issue, I'd like to get member deletion in soon!)

**Comment by cesy** [comment]
Happy to go for the default scope, just don't have time to code the rest of this anytime soon. Very happy for someone else to pick it up and run with it, given the spam.

**Comment by maco** [comment]
i'm picking this up

**Comment by pozorvlak** [comment]
@maco: thank you!

File: app/controllers/registrations_controller.rb [hunk 2]
```diff
@@ -42,5 +46,89 @@ def update
...
+  params[:member][:password_confirmation].present? ||
+  params[:member][:deleted].present?
+end
+
+# marking someone as deleted has various effects
+def markdelete
+
+    # move any of their crops to cropbot
+    if Role.crop_wranglers && (Role.crop_wranglers.include? @member)
+      cropbot = Member.find_by_login_name('ex_wrangler')
```
Comment (pozorvlak) [inline app/controllers/registrations_controller.rb]: `ex_wrangler` should be `cropbot`, I think.

File: app/controllers/registrations_controller.rb [hunk 3]
```diff
@@ -42,5 +46,89 @@ def update
...
+          crop.save!
+        end
+      end
+    end
+
+    # mark their comments as deleted
+    ex_member = Member.find_by_login_name('ex_member')
+    if Comment.find_by(author: @member)
+      Comment.where(author: @member).each do |comment|
+        Comment.update(comment, author: ex_member, body: "This comment was removed as the author deleted their account.")
```
Comment (pozorvlak) [inline app/controllers/registrations_controller.rb]: I'm not convinced this is the best approach. It preserves threading, but it's still destructive - we can't undo the deletion without restoring from backup. Alternatively, we could make comments/posts/gardens/harvests/etc soft-deletable, and render soft-deleted comments as "This comment was removed...". That would allow us to translate the "this comment was removed..." message. On the other hand:

 - it would be more work
 - it would still be destructive: delete a comment, then delete your account. Then undelete your account. Should we undelete the comment?

Maybe we could scope all gardens/comments/etc to "owner has not been soft-deleted"? I dunno. I'm tired.

**Comment by maco** [comment]
I agree @pozorvlak. Yes, to soft-delete. Yes, to the translatable
soft-render thing you said. When I finish figuring out the scope stuff,
I'll look at that.

On Wed, Feb 15, 2017 at 3:54 PM pozorvlak <notifications@github.com> wrote:

*@pozorvlak* commented on this pull request.
------------------------------

In app/controllers/registrations_controller.rb
<https://github.com/Growstuff/growstuff/pull/743#pullrequestreview-22107330>
:

> +    if Role.crop_wranglers && (Role.crop_wranglers.include? @member)
+      cropbot = Member.find_by_login_name('ex_wrangler')
+      if Crop.find_by(creator: @member)
+        # this is ugly, need to make it more efficient
+        Crop.where(creator: @member).each do |crop|
+          Crop.update(crop, creator: cropbot)
+          crop.save!
+        end
+      end
+    end
+
+    # mark their comments as deleted
+    ex_member = Member.find_by_login_name('ex_member')
+    if Comment.find_by(author: @member)
+      Comment.where(author: @member).each do |comment|
+        Comment.update(comment, author: ex_member, body: "This
comment was removed as the author deleted their account.")

I'm not convinced this is the best approach. It preserves threading, but
it's still destructive - we can't undo the deletion without restoring from
backup. Alternatively, we could make comments/posts/gardens/harvests/etc
soft-deletable, and render soft-deleted comments as "This comment was
removed...". That would allow us to translate the "this comment was
removed..." message. On the other hand:

   - it would be more work
   - it would still be destructive: delete a comment, then delete your
   account. Then undelete your account. Should we undelete the comment?

Maybe we could scope all gardens/comments/etc to "owner has not been
soft-deleted"? I dunno. I'm tired.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/Growstuff/growstuff/pull/743#pullrequestreview-22107330>,
or mute the thread

File: app/views/home/index.html.haml [hunk 4]
```diff
@@ -2,6 +2,9 @@
   .col-md-12
     - if member_signed_in?
       %h1= t('.welcome', site_name: ENV['GROWSTUFF_SITE_NAME'], member_name: current_member)
+      // TODO: remove this change before going live, just makes it easier to spot issues while testing
+      - if current_member.deleted?
+        %h1= "You are deleted and something is wrong"
```
Comment (Br3nda) [inline app/views/home/index.html.haml]: can we call t() ? the rest of the file mostly as this.

File: config/routes.rb [hunk 5]
```diff
@@ -9,6 +9,7 @@
     get '/members/unsubscribe/:message' => 'members#unsubscribe', :as => 'unsubscribe_member'
   end
   match '/members/:id/finish_signup' => 'members#finish_signup', via: [:get, :patch], :as => :finish_signup
+  match 'members/:id' => 'members#delete', :via => :delete
```
Comment (Br3nda) [inline config/routes.rb]: /members is getting quite full

how about instead
```
namespace :admin do
  resources :members
end
```
Then we can have a separate controller and keep the admin versus normal membership stuff apart.

**Comment by Br3nda** [comment]
Things we want to hide:
Member profile
Gardens info 
Plantings info
Harvests info
Seeds info


Things we want to keep showing:
Plant part stats (what a crop was harvested for) 
Use their plantings data in predictions of other plantings maturity date. 

Things I'm not sure of:
Photos. If they are CC licensed, we should show them on crops still.

**Comment by cesy** [comment]
Closing as superseded by #1341.
